amaranth_soc/wishbone/sram.py

```diff
@@ -0,0 +1,113 @@
+from amaranth import *
+from amaranth.lib import wiring
+from amaranth.lib.wiring import In
+from amaranth.lib.memory import MemoryData, Memory
+from amaranth.utils import exact_log2
+
+from ..memory import MemoryMap
+from .bus import Signature
+
+
+__all__ = ["WishboneSRAM"]
+
+
+class WishboneSRAM(wiring.Component):
+    """Wishbone-attached SRAM.
+
+    Wishbone bus accesses have a latency of one clock cycle. Incremental and constant address
+    bursts are supported.
+
+    Arguments
+    ---------
+    size : :class:`int`, power of two
+        SRAM size, in units of ``granularity`` bits.
+    data_width : ``8``, ``16``, ``32`` or ``64``
+        Wishbone bus data width.
+    granularity : ``8``, ``16``, ``32`` or ``64``, optional
+        Wishbone bus granularity. If unspecified, it defaults to ``data_width``.
+    writable : bool
+        Write capability. If disabled, writes are ignored. Enabled by default.
+    init : iterable of initial values, optional
+        Initial values for memory rows. There are ``(size * granularity) // data_width`` rows,
+        and each row has a shape of ``unsigned(data_width)``.
+
+    Members
+    -------
+    wb_bus : ``In(wishbone.Signature(...))``
+        Wishbone bus interface.
+
+    Raises
+    ------
+    :exc:`ValueError`
+        If ``size * granularity`` is lesser than ``data_width``.
+    """
+    def __init__(self, *, size, data_width, granularity=None, writable=True, init=()):
+        if granularity is None:
+            granularity = data_width
+
+        if not isinstance(size, int) or size <= 0 or size & size-1:
+            raise TypeError(f"Size must be an integer power of two, not {size!r}")
+        if data_width not in (8, 16, 32, 64):
+            raise TypeError(f"Data width must be 8, 16, 32 or 64, not {data_width!r}")
+        if granularity not in (8, 16, 32, 64):
+            raise TypeError(f"Granularity must be 8, 16, 32 or 64, not {granularity!r}")
+        if size * granularity < data_width:
+            raise ValueError(f"The product of size {size} and granularity {granularity} must be "
+                             f"greater than or equal to data width {data_width}, not "
+                             f"{size * granularity}")
+
+        self._size     = size
+        self._writable = bool(writable)
+        self._mem_data = MemoryData(depth=(size * granularity) // data_width,
+                                    shape=unsigned(data_width), init=init)
+        self._mem      = Memory(self._mem_data)
+
+        super().__init__({"wb_bus": In(Signature(addr_width=exact_log2(self._mem.depth),
+                                                 data_width=data_width, granularity=granularity,
+                                                 features=("cti", "bte")))})
```

Comment:
I think these need to be configurable or you can't connect the memory to a bus without CTI/BTE.

Reply:
Wishbone interfaces are currently not `connect()`ed in Amaranth SoC, and compatiblity is handled by the [primitives themselves](https://github.com/amaranth-lang/amaranth-soc/blob/058057f38b18537585a02e1db7955a184a1c5334/amaranth_soc/wishbone/bus.py#L361-L366), with interoperable fallbacks. I think we can make this a guideline for future Wishbone primitives.

Reply:
To clarify, I have a preference for not making it configurable.

Reply:
Ah, I forgot about this. We don't need to touch it until it becomes a problem, at least, not at my insistence.

Reply:
On the other hand, `WishboneSRAM` support for ERR or RTY would have to be configurable, as they can be used to terminate bus transactions.